public/app/plugins/datasource/influxdb/components/ConfigEditor.tsx

```diff
@@ -0,0 +1,223 @@
+import React, { PureComponent, ChangeEvent } from 'react';
+import { DataSourcePluginOptionsEditorProps, DataSourceSettings, SelectableValue } from '@grafana/data';
+import { DataSourceHttpSettings, FormLabel, Input, SecretFormField, Select } from '@grafana/ui';
+import { InfluxOptions, InfluxSecureJsonData } from '../types';
+
+export type Props = DataSourcePluginOptionsEditorProps<InfluxOptions>;
+
+type InfluxDataSourceSettings = DataSourceSettings<InfluxOptions, InfluxSecureJsonData>;
+
+export interface State {
+  config: InfluxDataSourceSettings;
+  httpModes: SelectableValue[];
+}
+
+export class ConfigEditor extends PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    const { options } = this.props;
+
+    this.state = {
+      config: ConfigEditor.defaults(options),
+      httpModes: [{ label: 'GET', value: 'GET' }, { label: 'POST', value: 'POST' }],
+    };
+  }
+
+  static getDerivedStateFromProps(props: Props, state: State) {
+    return {
+      ...state,
+      config: ConfigEditor.defaults(props.options),
+    };
+  }
+
+  static defaults = (options: any) => {
+    options.secureJsonData = options.secureJsonData || {};
+    options.secureJsonData.password = options.secureJsonData.password || '';
+    return options;
+  };
+
+  updateDatasource = async (config: any) => {
```

Comment:
any?   can you add some comments explaining what is happening here, and why?

Reply:
If you put the type here, it won't allow for string index use but can add comments.

Reply:
hymm -- that should be a red flag for fishy business!  Is the code below really required?  I would expect `config.secureJsonFields` to be based on what props are saved... BUT since this caches everything in the state it needs to do that itself.

Check [BasicAuthSettings.tsx](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/DataSourceSettings/BasicAuthSettings.tsx)... and we follow that pattern?  Can we just use that component to set the user/pass directly?

no need to put `config: InfluxDataSourceSettings;` in state either... I think that leaves us with an empty State object!

Reply:
IIUC, `secureJsonData` only needs to be sent when you are changing the field -- it will keep any existing values unless you send an object.  `secureJsonFields` is returned from the backend and just flags what is set -- you should never need to manage it directly.

Can you take a stab at updating some comments on:
https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/datasource.ts#L497
that match how you see it behave?

Reply:
Well, I do this because on Azure, if you send an empty string to the update, it will erase the secrets on the backend but since this backend on this plugin won't allow you remove any values, it's not needed.